man/fabric.7.md

```diff
@@ -238,6 +238,65 @@ FI_LOG_SUBSYS environment variables.
 - *mr*
 : Provides output specific to memory registration.
 
+# FILE SYSTEM CONFIGRATION INTERFACE
+
+Libfabric defines a set of file system interfaces that may be used to
+set global or provider specific parameters.  Conceptually, this interface
+is similar to the /proc interfaces used by ip(7) and tcp(7).
+Parameters can be accessed by reading or writing files in
+%SYSCONFDIR%/rdma/fabric.  The files are typically found at
+/etc/rdma/fabric or /usr/local/etc/rdma/fabric.
+The exact location of the files is controllable through build parameters,
+but for simplicity, the following documentation assumes that they are found
+under /etc/rdma/fabric.  The files do not exist by default and must be
+created by the administrator or Linux distro.  
+
+Parameter files located under /etc/rdma/fabric are treated as global and
+apply to all relevant providers.  Provider specific parameters may be set
+by writing to a file by the same name, but in a provider specific
+subdirectory.  For example, parameter files under /etc/rdma/fabric/verbs
+will apply only to the verbs provider.
+
+Note that providers may select alternate values that those
+specified.  This may be necessary in order to adhere to memory alignment
+restrictions, hardware maximums, or protocol limitations, for example.
+
+Most files are associated with a single value, with the data type
+shown in the descriptions below.  Administrators can set and get the values
+using a method similar to this example:
+
+% mkdir /etc/rdma
+% mkdir /etc/rdma/fabric
+% echo 64 > /etc/rdma/fabric/def_inject_size
+% cat /etc/rdma/fabric/def_inject_size
+  64
+
+In some cases, system files may conflict with libfabric or provider
+environment variables.  In most cases, if both values are provided, the
+environment variable takes precedence.  The exception are that parameter
+files that specify a maximum will limit environment variable settings.
```

Comment:
I think param files are fine, and Env vars are fine. But the usage model needs to be clear and unambiguous.

E.G., open MPI supports multiple methods of setting config vars. But *all* config vars are settable through *all* methods (e.g., file, Env var, etc). That way, everything is equal.

There's a clear precedence set in open MPI for setting these vars: files are lowest priority (system file, then user file), then Env vars, then the mpirun command line is the highest priority (but that isn't relevant here, obviously).

Sent from my phone. No type good.

On Oct 30, 2015, at 5:28 PM, Sean Hefty <notifications@github.com<mailto:notifications@github.com>> wrote:


In man/fabric.7.md<https://github.com/ofiwg/libfabric/pull/1411#discussion_r43554510>:

> +restrictions, hardware maximums, or protocol limitations, for example.
> +
> +Most files are associated with a single value, with the data type
> +shown in the descriptions below.  Administrators can set and get the values
> +using a method similar to this example:
> +
> +% mkdir /etc/rdma
> +% mkdir /etc/rdma/fabric
> +% echo 64 > /etc/rdma/fabric/def_inject_size
> +% cat /etc/rdma/fabric/def_inject_size
> +  64
> +
> +In some cases, system files may conflict with libfabric or provider
> +environment variables.  In most cases, if both values are provided, the
> +environment variable takes precedence.  The exception are that parameter
> +files that specify a maximum will limit environment variable settings.


I thought about this, but which would take precedence? IMO it should be the parameter files, since those can be secure. My concern was that a developer setting an environment variable would see it being ignored. If we're okay with that (maybe displaying a log message), then I will update the man page. But first, are we okay with the entire approach of having parameter files?

—
Reply to this email directly or view it on GitHub<https://github.com/ofiwg/libfabric/pull/1411/files#r43554510>.